commit e3c2af5d49bb96ad460ef6aa873975fb0f7c7f17
Author: Justin Stitt <justinstitt@google.com>
Date:   Fri Oct 20 18:59:34 2023 +0000

    power: supply: bq2515x: replace deprecated strncpy with strscpy
    
    strncpy() is deprecated for use on NUL-terminated destination strings
    [1] and as such we should prefer more robust and less ambiguous string
    interfaces.
    
    We expect bq2515x->model_name to be NUL-terminated based on its usage with
    sysfs_emit and format strings:
    
    val->strval is assigned to bq2515x->model_name in
    bq2515x_mains_get_property():
    |       val->strval = bq2515x->model_name;
    
    ... then in power_supply_sysfs.c we use value.strval with a format string:
    |       ret = sysfs_emit(buf, "%s\n", value.strval);
    
    we assigned value.strval via:
    |       ret = power_supply_get_property(psy, psp, &value);
    ... which invokes psy->desc->get_property():
    |       return psy->desc->get_property(psy, psp, val);
    
    with bq2515x_mains_get_property():
    |       static const struct power_supply_desc bq2515x_mains_desc = {
    ...
    |               .get_property           = bq2515x_mains_get_property,
    
    Moreover, no NUL-padding is required as bq2515x is zero-allocated in
    bq2515x_charger.c:
    |       bq2515x = devm_kzalloc(dev, sizeof(*bq2515x), GFP_KERNEL);
    
    Considering the above, a suitable replacement is `strscpy` [2] due to
    the fact that it guarantees NUL-termination on the destination buffer
    without unnecessarily NUL-padding.
    
    Let's also opt to use the more idiomatic strscpy() usage of (dest, src,
    sizeof(dest)) as this more closely ties the destination buffer and the
    length.
    
    Link: https://www.kernel.org/doc/html/latest/process/deprecated.html#strncpy-on-nul-terminated-strings [1]
    Link: https://manpages.debian.org/testing/linux-manual-4.8/strscpy.9.en.html [2]
    Link: https://github.com/KSPP/linux/issues/90
    Cc: linux-hardening@vger.kernel.org
    Signed-off-by: Justin Stitt <justinstitt@google.com>
    Similar-to: https://lore.kernel.org/all/20231020-strncpy-drivers-power-supply-bq24190_charger-c-v1-1-e896223cb795@google.com/
    Reviewed-by: Kees Cook <keescook@chromium.org>
    Link: https://lore.kernel.org/r/20231020-strncpy-drivers-power-supply-bq2515x_charger-c-v1-1-46664c6edf78@google.com
    Signed-off-by: Sebastian Reichel <sebastian.reichel@collabora.com>

diff --git a/drivers/power/supply/bq2515x_charger.c b/drivers/power/supply/bq2515x_charger.c
index 49fa6386a509..a3424f67f2b1 100644
--- a/drivers/power/supply/bq2515x_charger.c
+++ b/drivers/power/supply/bq2515x_charger.c
@@ -1089,7 +1089,7 @@ static int bq2515x_probe(struct i2c_client *client)
 
 	bq2515x->dev = dev;
 
-	strncpy(bq2515x->model_name, id->name, I2C_NAME_SIZE);
+	strscpy(bq2515x->model_name, id->name, sizeof(bq2515x->model_name));
 
 	bq2515x->info = i2c_get_match_data(client);
 	bq2515x->regmap = devm_regmap_init_i2c(client,